pkg/globalnet/controllers/gateway_monitor.go

```diff
@@ -184,25 +185,18 @@ func (g *gatewayMonitor) handleCreatedOrUpdatedEndpoint(obj runtime.Object, numR
 
 		configureTCPMTUProbe()
 
-		g.syncMutex.Lock()
-		if !g.isGatewayNode {
-			g.isGatewayNode = true
-
+		if g.isGatewayNode.CompareAndSwap(false, true) {
 			err := g.startControllers()
```

Comment:
I think this is really what needs to be protected — `stopControllers()` mustn’t be called while `startControllers()` is in progress (_i.e._ `controllers` needs to be synced).

Reply:
yeah I believe that's correct. `Stop` is called outside the informer thread otherwise we really wouldn't need any synchronization.